File: go/analyze/rules.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package analyze

import (
	"context"
	"regexp"
	"strings"

	keystonev1alpha1 "github.com/dogukanturhal/keystone/api/v1alpha1"
)

// Phase 11.2 analyzers. Each one is 10-30 LOC — intentionally simple so
// reviewers can audit them easily. Complex rules that need SQL parsing
// (e.g. matching column in ALTER context) are deferred to Phase 11.3
// where we introduce a minimal SQL tokenizer.

// statementMatches returns a finding per file whose content matches re.
func statementMatches(m *Migration, re *regexp.Regexp, sev keystonev1alpha1.LintLevel, rule, msg string) []Finding {
	var out []Finding
	for _, f := range m.Files {
		for i, line := range strings.Split(f.Body, "\n") {
			if re.MatchString(line) {
				out = append(out, Finding{
					Rule:     rule,
					Severity: sev,
					File:     f.Name,
					Line:     int32(i + 1),
					Message:  msg,
				})
			}
		}
	}
	return out
}

// -- Rule: no-drop-table ------------------------------------------------

type NoDropTable struct{}

func (NoDropTable) ID() string { return "no-drop-table" }
func (NoDropTable) Description() string {
	return "refuse DROP TABLE; use soft-delete or retention policy"
}

var reDropTable = regexp.MustCompile(`(?i)\bDROP\s+TABLE\b`)

func (a *NoDropTable) Check(_ context.Context, m *Migration) ([]Finding, error) {
	return statementMatches(m, reDropTable, keystonev1alpha1.LintLevelError,
		a.ID(),
		"DROP TABLE removes data irrecoverably; use a soft-delete column or "+
			"schedule via a separate destructive-migration workflow with operator sign-off"), nil
}

// -- Rule: no-drop-column -----------------------------------------------

type NoDropColumn struct{}

func (NoDropColumn) ID() string { return "no-drop-column" }
func (NoDropColumn) Description() string {
	return "prefer pgroll-expand-contract drop_column over ALTER TABLE DROP COLUMN"
}

var reDropColumn = regexp.MustCompile(`(?i)\bALTER\s+TABLE\b[^;]*\bDROP\s+COLUMN\b`)

func (a *NoDropColumn) Check(_ context.Context, m *Migration) ([]Finding, error) {
	return statementMatches(m, reDropColumn, keystonev1alpha1.LintLevelError,
		a.ID(),
		"ALTER TABLE DROP COLUMN is data-destructive and takes AccessExclusive lock. "+
			"Use strategy=pgroll-expand-contract with kind=drop_column for the two-phase safe drop"), nil
}

// -- Rule: prefer-concurrent-index-creation -----------------------------

type PreferConcurrentIndexCreation struct{}

func (PreferConcurrentIndexCreation) ID() string { return "prefer-concurrent-index-creation" }
func (PreferConcurrentIndexCreation) Description() string {
	return "CREATE INDEX takes ShareLock for the duration; add CONCURRENTLY for zero-downtime"
}

var reCreateIndexConcurrently = regexp.MustCompile(`(?i)\bCREATE\s+(UNIQUE\s+)?INDEX\s+CONCURRENTLY\b`)
var reCreateIndex = regexp.MustCompile(`(?i)\bCREATE\s+(UNIQUE\s+)?INDEX\b`)

func (a *PreferConcurrentIndexCreation) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		for i, line := range strings.Split(f.Body, "\n") {
			if reCreateIndex.MatchString(line) && !reCreateIndexConcurrently.MatchString(line) {
				out = append(out, Finding{
					Rule:     a.ID(),
					Severity: keystonev1alpha1.LintLevelWarning,
					File:     f.Name,
					Line:     int32(i + 1),
					Message: "CREATE INDEX without CONCURRENTLY locks writes for the scan. " +
						"Add CONCURRENTLY unless the table is empty or the migration is in a maintenance window.",
				})
			}
		}
	}
	return out, nil
}

// -- Rule: require-primary-key ------------------------------------------

type RequirePrimaryKey struct{}

func (RequirePrimaryKey) ID() string { return "require-primary-key" }
func (RequirePrimaryKey) Description() string {
	return "every CREATE TABLE must declare a PRIMARY KEY"
}

var reCreateTable = regexp.MustCompile(`(?i)\bCREATE\s+TABLE\b`)
var rePrimaryKey = regexp.MustCompile(`(?i)\bPRIMARY\s+KEY\b`)

func (a *RequirePrimaryKey) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		// Crude: CREATE TABLE without PRIMARY KEY somewhere in the
		// same file. False negatives (multi-stmt files) and positives
		// (comments) are tolerated — this is a warning, not a block.
		if reCreateTable.MatchString(f.Body) && !rePrimaryKey.MatchString(f.Body) {
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelWarning,
				File:     f.Name,
				Message: "file contains CREATE TABLE but no PRIMARY KEY clause detected. " +
					"Tables without a PK can't be replicated by logical replication or migrated via pgroll.",
			})
		}
	}
	return out, nil
}

// -- Rule: no-alter-column-type-in-place --------------------------------

type NoAlterColumnTypeInPlace struct{}

func (NoAlterColumnTypeInPlace) ID() string { return "no-alter-column-type-in-place" }
func (NoAlterColumnTypeInPlace) Description() string {
	return "ALTER COLUMN ... TYPE rewrites the table; use pgroll alter_column_type"
}

var reAlterColumnType = regexp.MustCompile(`(?i)\bALTER\s+COLUMN\s+\w+\s+TYPE\b|\bALTER\s+COLUMN\s+\w+\s+SET\s+DATA\s+TYPE\b`)

func (a *NoAlterColumnTypeInPlace) Check(_ context.Context, m *Migration) ([]Finding, error) {
	return statementMatches(m, reAlterColumnType, keystonev1alpha1.LintLevelError,
		a.ID(),
		"ALTER COLUMN TYPE rewrites the table and takes AccessExclusive lock. "+
			"Use strategy=pgroll-expand-contract with kind=alter_column_type for zero-downtime type changes"), nil
}

// -- Rule: no-add-required-field-without-default ------------------------

type NoAddRequiredFieldWithoutDefault struct{}

func (NoAddRequiredFieldWithoutDefault) ID() string { return "no-add-required-field-without-default" }
func (NoAddRequiredFieldWithoutDefault) Description() string {
	return "ADD COLUMN NOT NULL without DEFAULT rewrites every row; add default or allow NULL"
}

var reAddColumnNotNullNoDefault = regexp.MustCompile(`(?i)\bADD\s+COLUMN\s+\w+\s+[A-Za-z]+[^;]*\bNOT\s+NULL\b`)
var reHasDefault = regexp.MustCompile(`(?i)\bDEFAULT\s+\S+`)

func (a *NoAddRequiredFieldWithoutDefault) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		for i, line := range strings.Split(f.Body, "\n") {
			if reAddColumnNotNullNoDefault.MatchString(line) && !reHasDefault.MatchString(line) {
				out = append(out, Finding{
					Rule:     a.ID(),
					Severity: keystonev1alpha1.LintLevelError,
					File:     f.Name,
					Line:     int32(i + 1),
					Message: "ADD COLUMN ... NOT NULL without DEFAULT forces a full-table rewrite. " +
						"Either add a DEFAULT (PG 11+ instant add) or use kind=add_column with nullable=true + enforceNotNullInContract",
				})
			}
		}
	}
	return out, nil
}

// -- Rule: no-truncate --------------------------------------------------

type NoTruncate struct{}

func (NoTruncate) ID() string { return "no-truncate" }
func (NoTruncate) Description() string {
	return "TRUNCATE is data-destructive and bypasses DELETE triggers"
}

var reTruncate = regexp.MustCompile(`(?i)\bTRUNCATE\b`)

func (a *NoTruncate) Check(_ context.Context, m *Migration) ([]Finding, error) {
	// Gate: only fire when the containing statement IS a TRUNCATE. The
	// raw `\bTRUNCATE\b` match would otherwise trip on a privilege name
	// in `GRANT SELECT, INSERT, UPDATE, DELETE, TRUNCATE ON TABLE …`
	// (verified live 2026-05-12 on example-service migration 234).
	var out []Finding
	for _, f := range m.Files {
		for _, loc := range reTruncate.FindAllStringIndex(f.Body, -1) {
			if !stmtVerbIs(f.Body, loc[0], "TRUNCATE") {
				continue
			}
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelError,
				File:     f.Name,
				Line:     int32(lineOfOffset(f.Body, loc[0])),
				Message: "TRUNCATE removes all rows irrecoverably and bypasses ON DELETE triggers. " +
					"Use DELETE if you need row-level side effects, or policy-override=true if truly intended",
			})
		}
	}
	return out, nil
}

// -- Rule: no-grant-all -------------------------------------------------

type NoGrantAll struct{}

func (NoGrantAll) ID() string { return "no-grant-all" }
func (NoGrantAll) Description() string {
	return "GRANT ALL is a privilege-escalation footgun; enumerate the privileges you actually need"
}

var reGrantAll = regexp.MustCompile(`(?i)\bGRANT\s+ALL\b`)

func (a *NoGrantAll) Check(_ context.Context, m *Migration) ([]Finding, error) {
	return statementMatches(m, reGrantAll, keystonev1alpha1.LintLevelWarning,
		a.ID(),
		"GRANT ALL includes future privileges added by PG upgrades. "+
			"Prefer an explicit privilege list: GRANT SELECT, INSERT, UPDATE, DELETE ..."), nil
}

// -- Rule: require-statement-timeout-on-ddl -----------------------------

type RequireStatementTimeoutOnDDL struct{}

func (RequireStatementTimeoutOnDDL) ID() string { return "require-statement-timeout-on-ddl" }
func (RequireStatementTimeoutOnDDL) Description() string {
	return "bare ALTER TABLE without statement_timeout can block forever on lock contention"
}

var reDDL = regexp.MustCompile(`(?i)\b(ALTER|CREATE|DROP)\s+(TABLE|INDEX|CONSTRAINT)\b`)
var reStatementTimeout = regexp.MustCompile(`(?i)\bSET\s+(LOCAL\s+)?statement_timeout\b`)

func (a *RequireStatementTimeoutOnDDL) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		if reDDL.MatchString(f.Body) && !reStatementTimeout.MatchString(f.Body) {
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelNotice,
				File:     f.Name,
				Message: "DDL migration without SET LOCAL statement_timeout. " +
					"Keystone's runner sets a 30s default at pool level, but explicit SET in the migration " +
					"makes the behaviour visible to reviewers.",
			})
		}
	}
	return out, nil
}

// -- Rule: no-transaction-around-concurrent-index ------------------------

type NoTransactionAroundConcurrentIndex struct{}

func (NoTransactionAroundConcurrentIndex) ID() string {
	return "no-transaction-around-concurrent-index"
}
func (NoTransactionAroundConcurrentIndex) Description() string {
	return "CREATE INDEX CONCURRENTLY cannot run inside a transaction block"
}

// reBeginStatement matches an explicit top-level transaction-control BEGIN
// (with or without TRANSACTION/WORK qualifier). Anchored at start-of-line
// after optional whitespace, terminated by `;` or end-of-line. Critically
// this does NOT match PL/pgSQL block-opener BEGIN — which is bare `BEGIN`
// inside a `DO $$ ... $$;` block or function body and has no terminator on
// the same line. Using `(?im)` so `^` anchors per-line and the match is
// case-insensitive.
var reBeginStatement = regexp.MustCompile(`(?im)^\s*BEGIN\s*(?:TRANSACTION|WORK)?\s*;`)

// reDoBlock matches a top-level DO $tag$ ... $tag$; block (where $tag$ is
// either `$$` or a custom dollar-quoted tag like `$func$`). Stripped from
// migration bodies before scanning for transaction-control BEGIN, so the
// PL/pgSQL `BEGIN`/`END` block delimiters inside DO blocks don't confuse
// transaction-control detection. Differ emits idempotent FK creation as
// `DO $$ BEGIN IF NOT EXISTS (...) THEN ALTER TABLE ... END IF; END $$;` —
// without this strip, every such bundle would false-positive
// no-transaction-around-concurrent-index. (?s) makes . span newlines.
//
// Note: Go's regexp (RE2) doesn't support backreferences, so we accept any
// dollar-quoted close tag, not strictly the same one as the open tag. For
// the linter's purpose (avoiding false positives on PL/pgSQL BEGIN) the
// approximation is harmless — in practice the differ emits balanced
// `$$ ... $$` blocks and never nests them.
var reDoBlock = regexp.MustCompile(`(?is)\bDO\s+\$\w*\$.*?\$\w*\$\s*;`)

func (a *NoTransactionAroundConcurrentIndex) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		// Strip DO blocks before checking for transaction-control BEGIN.
		// PL/pgSQL `BEGIN` inside `DO $$ ... $$;` is a block opener, not
		// transaction control, and Postgres tolerates it inside the
		// runner's per-file tx. Only EXPLICIT top-level BEGIN/COMMIT
		// would actually break CONCURRENTLY at runtime.
		bodyNoDo := reDoBlock.ReplaceAllString(f.Body, "")
		if reCreateIndexConcurrently.MatchString(bodyNoDo) && reBeginStatement.MatchString(bodyNoDo) {
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelError,
				File:     f.Name,
				Message: "CREATE INDEX CONCURRENTLY cannot run inside a transaction. " +
					"Move the CONCURRENTLY statement into its own file (Keystone's runner applies each file in its own tx).",
			})
		}
	}
	return out, nil
}

File: rust/keystone-sdk/src/analyze/rules.rs
// SPDX-License-Identifier: Apache-2.0

//! Phase 11.2 base pack — structural safety. Ported from `rules.go`.

use std::sync::LazyLock;

use regex::Regex;

use super::{
    finding, line_of_offset, statement_matches, stmt_verb_is, Analyzer, Finding, LintLevel,
    Migration,
};

macro_rules! re {
    ($name:ident, $pat:expr) => {
        static $name: LazyLock<Regex> = LazyLock::new(|| Regex::new($pat).expect($pat));
    };
}

re!(RE_DROP_TABLE, r"(?i)\bDROP\s+TABLE\b");
re!(
    RE_DROP_COLUMN,
    r"(?i)\bALTER\s+TABLE\b[^;]*\bDROP\s+COLUMN\b"
);
re!(
    RE_CREATE_INDEX_CONCURRENTLY,
    r"(?i)\bCREATE\s+(UNIQUE\s+)?INDEX\s+CONCURRENTLY\b"
);
re!(RE_CREATE_INDEX, r"(?i)\bCREATE\s+(UNIQUE\s+)?INDEX\b");
re!(RE_CREATE_TABLE, r"(?i)\bCREATE\s+TABLE\b");
re!(RE_PRIMARY_KEY, r"(?i)\bPRIMARY\s+KEY\b");
re!(
    RE_ALTER_COLUMN_TYPE,
    r"(?i)\bALTER\s+COLUMN\s+\w+\s+TYPE\b|\bALTER\s+COLUMN\s+\w+\s+SET\s+DATA\s+TYPE\b"
);
re!(
    RE_ADD_COLUMN_NOT_NULL_NO_DEFAULT,
    r"(?i)\bADD\s+COLUMN\s+\w+\s+[A-Za-z]+[^;]*\bNOT\s+NULL\b"
);
re!(RE_HAS_DEFAULT, r"(?i)\bDEFAULT\s+\S+");
re!(RE_TRUNCATE, r"(?i)\bTRUNCATE\b");
re!(RE_GRANT_ALL, r"(?i)\bGRANT\s+ALL\b");
re!(
    RE_DDL,
    r"(?i)\b(ALTER|CREATE|DROP)\s+(TABLE|INDEX|CONSTRAINT)\b"
);
re!(
    RE_STATEMENT_TIMEOUT,
    r"(?i)\bSET\s+(LOCAL\s+)?statement_timeout\b"
);
re!(
    RE_BEGIN_STATEMENT,
    r"(?im)^\s*BEGIN\s*(?:TRANSACTION|WORK)?\s*;"
);
re!(RE_DO_BLOCK, r"(?is)\bDO\s+\$\w*\$.*?\$\w*\$\s*;");

/// no-drop-table
pub struct NoDropTable;
impl Analyzer for NoDropTable {
    fn id(&self) -> &'static str {
        "no-drop-table"
    }
    fn description(&self) -> &'static str {
        "refuse DROP TABLE; use soft-delete or retention policy"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        statement_matches(
            m,
            &RE_DROP_TABLE,
            LintLevel::Error,
            self.id(),
            "DROP TABLE removes data irrecoverably; use a soft-delete column or \
             schedule via a separate destructive-migration workflow with operator sign-off",
        )
    }
}

/// no-drop-column
pub struct NoDropColumn;
impl Analyzer for NoDropColumn {
    fn id(&self) -> &'static str {
        "no-drop-column"
    }
    fn description(&self) -> &'static str {
        "prefer pgroll-expand-contract drop_column over ALTER TABLE DROP COLUMN"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        statement_matches(
            m,
            &RE_DROP_COLUMN,
            LintLevel::Error,
            self.id(),
            "ALTER TABLE DROP COLUMN is data-destructive and takes AccessExclusive lock. \
             Use strategy=pgroll-expand-contract with kind=drop_column for the two-phase safe drop",
        )
    }
}

/// prefer-concurrent-index-creation
pub struct PreferConcurrentIndexCreation;
impl Analyzer for PreferConcurrentIndexCreation {
    fn id(&self) -> &'static str {
        "prefer-concurrent-index-creation"
    }
    fn description(&self) -> &'static str {
        "CREATE INDEX takes ShareLock for the duration; add CONCURRENTLY for zero-downtime"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            for (i, line) in f.body.split('\n').enumerate() {
                if RE_CREATE_INDEX.is_match(line) && !RE_CREATE_INDEX_CONCURRENTLY.is_match(line) {
                    out.push(finding(
                        self.id(),
                        LintLevel::Warning,
                        &f.name,
                        (i + 1) as i32,
                        "CREATE INDEX without CONCURRENTLY locks writes for the scan. \
                         Add CONCURRENTLY unless the table is empty or the migration is in a maintenance window.",
                    ));
                }
            }
        }
        out
    }
}

/// require-primary-key
pub struct RequirePrimaryKey;
impl Analyzer for RequirePrimaryKey {
    fn id(&self) -> &'static str {
        "require-primary-key"
    }
    fn description(&self) -> &'static str {
        "every CREATE TABLE must declare a PRIMARY KEY"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            if RE_CREATE_TABLE.is_match(&f.body) && !RE_PRIMARY_KEY.is_match(&f.body) {
                out.push(finding(
                    self.id(),
                    LintLevel::Warning,
                    &f.name,
                    0,
                    "file contains CREATE TABLE but no PRIMARY KEY clause detected. \
                     Tables without a PK can't be replicated by logical replication or migrated via pgroll.",
                ));
            }
        }
        out
    }
}

/// no-alter-column-type-in-place
pub struct NoAlterColumnTypeInPlace;
impl Analyzer for NoAlterColumnTypeInPlace {
    fn id(&self) -> &'static str {
        "no-alter-column-type-in-place"
    }
    fn description(&self) -> &'static str {
        "ALTER COLUMN ... TYPE rewrites the table; use pgroll alter_column_type"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        statement_matches(
            m,
            &RE_ALTER_COLUMN_TYPE,
            LintLevel::Error,
            self.id(),
            "ALTER COLUMN TYPE rewrites the table and takes AccessExclusive lock. \
             Use strategy=pgroll-expand-contract with kind=alter_column_type for zero-downtime type changes",
        )
    }
}

/// no-add-required-field-without-default
pub struct NoAddRequiredFieldWithoutDefault;
impl Analyzer for NoAddRequiredFieldWithoutDefault {
    fn id(&self) -> &'static str {
        "no-add-required-field-without-default"
    }
    fn description(&self) -> &'static str {
        "ADD COLUMN NOT NULL without DEFAULT rewrites every row; add default or allow NULL"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            for (i, line) in f.body.split('\n').enumerate() {
                if RE_ADD_COLUMN_NOT_NULL_NO_DEFAULT.is_match(line)
                    && !RE_HAS_DEFAULT.is_match(line)
                {
                    out.push(finding(
                        self.id(),
                        LintLevel::Error,
                        &f.name,
                        (i + 1) as i32,
                        "ADD COLUMN ... NOT NULL without DEFAULT forces a full-table rewrite. \
                         Either add a DEFAULT (PG 11+ instant add) or use kind=add_column with nullable=true + enforceNotNullInContract",
                    ));
                }
            }
        }
        out
    }
}

/// no-truncate
pub struct NoTruncate;
impl Analyzer for NoTruncate {
    fn id(&self) -> &'static str {
        "no-truncate"
    }
    fn description(&self) -> &'static str {
        "TRUNCATE is data-destructive and bypasses DELETE triggers"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        // Gate: only fire when the containing statement IS a TRUNCATE (avoids
        // matching the privilege name in GRANT ... TRUNCATE ...).
        let mut out = Vec::new();
        for f in &m.files {
            for mat in RE_TRUNCATE.find_iter(&f.body) {
                if !stmt_verb_is(&f.body, mat.start(), "TRUNCATE") {
                    continue;
                }
                out.push(finding(
                    self.id(),
                    LintLevel::Error,
                    &f.name,
                    line_of_offset(&f.body, mat.start()),
                    "TRUNCATE removes all rows irrecoverably and bypasses ON DELETE triggers. \
                     Use DELETE if you need row-level side effects, or policy-override=true if truly intended",
                ));
            }
        }
        out
    }
}

/// no-grant-all
pub struct NoGrantAll;
impl Analyzer for NoGrantAll {
    fn id(&self) -> &'static str {
        "no-grant-all"
    }
    fn description(&self) -> &'static str {
        "GRANT ALL is a privilege-escalation footgun; enumerate the privileges you actually need"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        statement_matches(
            m,
            &RE_GRANT_ALL,
            LintLevel::Warning,
            self.id(),
            "GRANT ALL includes future privileges added by PG upgrades. \
             Prefer an explicit privilege list: GRANT SELECT, INSERT, UPDATE, DELETE ...",
        )
    }
}

/// require-statement-timeout-on-ddl
pub struct RequireStatementTimeoutOnDdl;
impl Analyzer for RequireStatementTimeoutOnDdl {
    fn id(&self) -> &'static str {
        "require-statement-timeout-on-ddl"
    }
    fn description(&self) -> &'static str {
        "bare ALTER TABLE without statement_timeout can block forever on lock contention"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            if RE_DDL.is_match(&f.body) && !RE_STATEMENT_TIMEOUT.is_match(&f.body) {
                out.push(finding(
                    self.id(),
                    LintLevel::Notice,
                    &f.name,
                    0,
                    "DDL migration without SET LOCAL statement_timeout. \
                     Keystone's runner sets a 30s default at pool level, but explicit SET in the migration \
                     makes the behaviour visible to reviewers.",
                ));
            }
        }
        out
    }
}

/// no-transaction-around-concurrent-index
pub struct NoTransactionAroundConcurrentIndex;
impl Analyzer for NoTransactionAroundConcurrentIndex {
    fn id(&self) -> &'static str {
        "no-transaction-around-concurrent-index"
    }
    fn description(&self) -> &'static str {
        "CREATE INDEX CONCURRENTLY cannot run inside a transaction block"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            // Strip DO blocks before checking for transaction-control BEGIN.
            let body_no_do = RE_DO_BLOCK.replace_all(&f.body, "");
            if RE_CREATE_INDEX_CONCURRENTLY.is_match(&body_no_do)
                && RE_BEGIN_STATEMENT.is_match(&body_no_do)
            {
                out.push(finding(
                    self.id(),
                    LintLevel::Error,
                    &f.name,
                    0,
                    "CREATE INDEX CONCURRENTLY cannot run inside a transaction. \
                     Move the CONCURRENTLY statement into its own file (Keystone's runner applies each file in its own tx).",
                ));
            }
        }
        out
    }
}

#[cfg(test)]
mod tests {
    use super::super::{by_rule, Migration};
    use super::*;

    fn mig(body: &str) -> Migration {
        Migration {
            files: vec![super::super::FileBody {
                name: "001.sql".into(),
                body: body.into(),
            }],
            ..Default::default()
        }
    }

    #[test]
    fn no_drop_table_fires() {
        assert_eq!(NoDropTable.check(&mig("DROP TABLE users;")).len(), 1);
        assert_eq!(NoDropTable.check(&mig("SELECT 1;")).len(), 0);
    }

    #[test]
    fn no_truncate_gated_by_verb() {
        // GRANT ... TRUNCATE must NOT fire (privilege name, not the verb).
        let f = NoTruncate.check(&mig("GRANT TRUNCATE ON t TO app;"));
        assert_eq!(f.len(), 0);
        // Real TRUNCATE fires.
        let f = NoTruncate.check(&mig("TRUNCATE t;"));
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].severity, LintLevel::Error);
    }

    #[test]
    fn prefer_concurrent_index_skips_concurrently() {
        assert_eq!(
            PreferConcurrentIndexCreation
                .check(&mig("CREATE INDEX i ON t (c);"))
                .len(),
            1
        );
        assert_eq!(
            PreferConcurrentIndexCreation
                .check(&mig("CREATE INDEX CONCURRENTLY i ON t (c);"))
                .len(),
            0
        );
    }

    #[test]
    fn require_primary_key_whole_file() {
        assert_eq!(
            RequirePrimaryKey
                .check(&mig("CREATE TABLE t (id int);"))
                .len(),
            1
        );
        assert_eq!(
            RequirePrimaryKey
                .check(&mig("CREATE TABLE t (id int PRIMARY KEY);"))
                .len(),
            0
        );
    }

    #[test]
    fn add_required_field_without_default() {
        let f = NoAddRequiredFieldWithoutDefault
            .check(&mig("ALTER TABLE t ADD COLUMN c int NOT NULL;"));
        assert_eq!(f.len(), 1);
        let f = NoAddRequiredFieldWithoutDefault
            .check(&mig("ALTER TABLE t ADD COLUMN c int NOT NULL DEFAULT 0;"));
        assert_eq!(f.len(), 0);
    }

    #[test]
    fn no_tx_around_concurrent_index_ignores_do_block() {
        // Explicit BEGIN; + CONCURRENTLY → fires.
        let f = NoTransactionAroundConcurrentIndex.check(&mig(
            "BEGIN;\nCREATE INDEX CONCURRENTLY i ON t (c);\nCOMMIT;",
        ));
        assert_eq!(f.len(), 1);
        // PL/pgSQL BEGIN inside a DO block must NOT trip it.
        let f = NoTransactionAroundConcurrentIndex.check(&mig(
            "DO $$ BEGIN PERFORM 1; END $$;\nCREATE INDEX CONCURRENTLY i ON t (c);",
        ));
        assert_eq!(f.len(), 0);
    }

    #[test]
    fn registry_smoke_counts() {
        // A migration that trips several rules at once.
        let m = mig("DROP TABLE old;\nCREATE INDEX i ON t (c);");
        let findings = NoDropTable.check(&m);
        let found = by_rule(&findings);
        assert_eq!(found.get("no-drop-table"), Some(&1));
    }
}
